### PR TITLE
#1945 Issue: update SingerHelper read method

### LINKED ISSUE #1945: Slack failing CI build
## Expected Behavior
Slack's CI build has been failing repeatedly. We should investigate and fix. See the [Connector health](https://docs.airbyte.io/integrations/connector-health) page for more information. 
```
2021-02-04 12:38:15 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - parsing
2021-02-04 12:38:15 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - executing command TestAction.TEARDOWN
2021-02-04 12:38:15 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - writing output
2021-02-04 12:38:15 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - complete

Failures (3):
  JUnit Jupiter:PythonTestSource:testIdenticalFullRefreshes()
    MethodSource [className = 'io.airbyte.integrations.standardtest.source.PythonTestSource', methodName = 'testIdenticalFullRefreshes', methodParameterTypes = '']
    => org.opentest4j.AssertionFailedError: Expected two full refresh syncs to produce the same records ==> expected: <449> but was: <448>
       org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
       org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
       org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
       org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:542)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.assertSameRecords(StandardSourceTest.java:445)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.testIdenticalFullRefreshes(StandardSourceTest.java:286)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
       java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       java.base/java.lang.reflect.Method.invoke(Method.java:564)
       [...]
  JUnit Jupiter:PythonTestSource:testIncrementalSyncWithState()
    MethodSource [className = 'io.airbyte.integrations.standardtest.source.PythonTestSource', methodName = 'testIncrementalSyncWithState', methodParameterTypes = '']
    => org.opentest4j.AssertionFailedError: Expected incremental sync to produce STATE messages ==> expected: <false> but was: <true>
       org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
       org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:40)
       org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:218)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.testIncrementalSyncWithState(StandardSourceTest.java:323)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
       java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       java.base/java.lang.reflect.Method.invoke(Method.java:564)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       [...]
  JUnit Jupiter:PythonTestSource:testEmptyStateIncrementalIdenticalToFullRefresh()
    MethodSource [className = 'io.airbyte.integrations.standardtest.source.PythonTestSource', methodName = 'testEmptyStateIncrementalIdenticalToFullRefresh', methodParameterTypes = '']
    => org.opentest4j.AssertionFailedError: Expected a full refresh sync and incremental sync with no input state to produce identical records ==> expected: <449> but was: <443>
       org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
       org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
       org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
       org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:542)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.assertSameRecords(StandardSourceTest.java:445)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.testEmptyStateIncrementalIdenticalToFullRefresh(StandardSourceTest.java:362)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
       java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       java.base/java.lang.reflect.Method.invoke(Method.java:564)
       [...]

Test run finished after 31742 ms
[         2 containers found      ]
[         0 containers skipped    ]
[         2 containers started    ]
[         0 containers aborted    ]
[         2 containers successful ]
[         0 containers failed     ]
[         7 tests found           ]
[         0 tests skipped         ]
[         7 tests started         ]
[         0 tests aborted         ]
[         4 tests successful      ]
[         3 tests failed          ]

There are failing tests. See https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/standard-source-tests for more information about the standard source test suite.


> Task :airbyte-integrations:connectors:source-slack-singer:standardSourceTestPython FAILED
FAILURE: Build failed with an exception.
```

- [ ] Check the correct operation of the Incremental for the connector.
- [ ] Check data for streams called during CI tests.

### LINKED ISSUE #2030: Google Analytics failing CI test.
## Current Behavior
*Found that the CI tests for Google Analytics pass through time. 
I looked at https://docs.airbyte.io/integrations/connector-health#sources and found that the problem is not new.
The issue is happening because two syncs don't produce exactly the same records.*
```
2021-02-10 12:27:56 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - parsing
2021-02-10 12:27:56 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - executing command TestAction.TEARDOWN
2021-02-10 12:27:56 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - writing output
2021-02-10 12:27:56 INFO i.a.c.i.LineGobbler(voidCall):69 - *** - complete

Failures (1):
  JUnit Jupiter:PythonTestSource:testIdenticalFullRefreshes()
    MethodSource [className = 'io.airbyte.integrations.standardtest.source.PythonTestSource', methodName = 'testIdenticalFullRefreshes', methodParameterTypes = '']
    => org.opentest4j.AssertionFailedError: Expected two full refresh syncs to produce the same records ==> expected: <1060> but was: <1061>
       org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
       org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
       org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
       org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:542)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.assertSameRecords(StandardSourceTest.java:445)
       io.airbyte.integrations.standardtest.source.StandardSourceTest.testIdenticalFullRefreshes(StandardSourceTest.java:286)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
       java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
       java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       java.base/java.lang.reflect.Method.invoke(Method.java:564)
       [...]

Test run finished after 34817 ms
[         2 containers found      ]
[         0 containers skipped    ]
[         2 containers started    ]
[         0 containers aborted    ]
[         2 containers successful ]
[         0 containers failed     ]
[         7 tests found           ]
[         0 tests skipped         ]
[         7 tests started         ]
[         0 tests aborted         ]
[         6 tests successful      ]
[         1 tests failed          ]

There are failing tests. See https://docs.airbyte.io/contributing-to-airbyte/building-new-connector/standard-source-tests for more information about the standard source test suite.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':airbyte-integrations:connectors:source-googleanalytics-singer:standardSourceTestPython'.
> Process 'command 'docker'' finished with non-zero exit value 1

```

## Expected Behavior
*CI tests should be passed every time.*

## Steps to Reproduce
1. Running CI testing for Google Analytics Source.

## Severity of the bug for you
High

### PR DESCRIPTION
**Pull request by yevhenii-ldv**
## What
*It was found that for all Singer-based connectors, the data read from the console can be not always quite correct. The problem was that we use two `IOTextWrappers`: one for `stdout`, and the second for `stderror`. In the code, we exit the loop and stop receiving data from the console as soon as one of the `IOTextWrappers` gives us an empty string (this means the end of the file). However, we need both `IOTextWrappers` to return an empty string and only then stop reading from the output console. Thus, the previous code sometimes exited the loop earlier and did not read a certain number of records and states, this was clearly seen on the "users" stream for the Slack connector..*

## How
*Describe the solution*

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `test.java`
1. `component.ts`
1. the rest

### CODE REVIEW

**Comment by yevhenii-ldv** [comment]
I left this PR as a WIP, as I want to work on it a little more and optimize it, removing such a large number of code nestings for method `read`.

**Review by sherifnada — APPROVED** [review]
Nice catch 👍

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 1]
```diff
@@ -163,48 +163,67 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
...
-                        else:
-                            logger.log_by_prefix(line, "INFO")
+            for std_data in SingerHelper._read_std_rows(p, is_message, transform):
+                if isinstance(std_data, AirbyteMessage):
+                    yield std_data
+                else:
+                    logger.log_by_prefix(*std_data)
+
+    @staticmethod
+    def _read_std_rows(process: subprocess.Popen, is_message, transform) -> Generator[AirbyteMessage, None, None]:
```
Comment (eugene-kulak) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:173]:
```suggestion
    def _read_std_rows(process: subprocess.Popen, is_message, transform) -> Iterator[AirbyteMessage]:
```

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 2]
```diff
@@ -163,48 +163,67 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
...
+            else:
+                # todo: check that messages match the discovered schema
+                stream_name = transformed_json["stream"]
+                out_record = AirbyteRecordMessage(
+                    stream=stream_name,
+                    data=transformed_json["record"],
+                    emitted_at=int(datetime.now().timestamp()) * 1000,
+                )
+                out_message = AirbyteMessage(type=Type.RECORD, record=out_record)
+                return transform(out_message)
```
Comment (eugene-kulak) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:226]: `return None`

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 3]
```diff
@@ -163,48 +163,67 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
     @staticmethod
     def read(logger, shell_command, is_message=(lambda x: True), transform=(lambda x: x)) -> Generator[AirbyteMessage, None, None]:
```
Comment (eugene-kulak) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:164]: please remove transform, I don't its usage it is not transparent enough why it is here

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 4]
```diff
@@ -163,48 +163,67 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
...
+                        if message_data is not None:
+                            yield message_data
                     else:
-                        logger.log_by_prefix(line, "ERROR")
+                        yield line, "INFO"
+                else:
+                    yield line, "ERROR"
+
+    @staticmethod
+    def _classify_and_convert_out_json_to_airbyte_message(out_json: Dict, transform) -> AirbyteMessage:
```
Comment (eugene-kulak) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:208]:
```suggestion
    def _airbyte_message_from_json(out_json: Mapping[str, Ayy], transform) -> Optional[AirbyteMessage]:
```

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 5]
```diff
@@ -161,50 +162,63 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
...
+            selects_list = sel.select()
+            empty_line_counter = 0
+            for key, _ in selects_list:
+                line = key.fileobj.readline()
+                if not line:
+                    empty_line_counter += 1
+                    if empty_line_counter >= len(selects_list):
+                        eof = True
+
+                    try:
```
Comment (sherifnada) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:195]: we should only run this section of code (the `try except` then `if process.returncode` etc.. if EOF has not been reached, so this should go outside the while loop

**Review by sherifnada — APPROVED** [review]
One comment about EOF, once it's addressed we can merge. Good catch! 🎉

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 6]
```diff
@@ -161,50 +162,63 @@ def get_catalogs(logger, shell_command: str, sync_mode_overrides: Dict[str, Sync
...
+                        raise Exception(f"Underlying command {process.args} is hanging")
+
+                    if process.returncode != 0:
+                        raise Exception(f"Underlying command {process.args} failed with exit code {process.returncode}")
+                else:
+                    yield line, key.fileobj
+
+    @staticmethod
+    def _airbyte_message_from_json(transformed_json: Mapping[str, Any]) -> Optional[AirbyteMessage]:
+        if transformed_json is None or transformed_json.get("type") == "SCHEMA" or transformed_json.get("type") == "ACTIVATE_VERSION":
```
Comment (eugene-kulak) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:207]: `if transformed_json is None or transformed_json.get("type") in {"SCHEMA", "ACTIVATE_VERSION"}:

**Pull request merged**